### ISSUE TITLE
Dexie.Syncable unhandledrejection

### ISSUE THREAD

**Issue by nevf** [dfahlander/Dexie.js#434]
@dfahlander 
I've read https://github.com/dfahlander/Dexie.js/issues/397 and am also extremelly interested in Dexie.Syncable with a Node.js backend and MongoDB. FYI I've spent some time evaluating PouchDB and have decided it isn't for me. 

Right now I am just fiddling around trying to grok how Syncable works. I'm using the ```WebSocketSyncProtocol.js``` and ```WebSocketSyncServer.js``` example code. 

If I don't have the server running ```db.syncable.connect(...).catch()``` catches on the first rejection, however subsequent ```Syncable -> doSync()``` calls raise an ```unhandledrejection``` which is I assume because ```doSync()``` calls itself via. setTimeout() and the subsequent rejection cannot be picked up by my code.

Further other than the ```promise.stack``` in my ```window.onunhandledrejection``` function I have no way to usefully determine why or who raised the rejection.

Related to this ```Dexie.Promise.on(..)``` is undefined ref:  https://github.com/dfahlander/Dexie.js/wiki/Promise.on.error which I though may have been helpfull.

The bottom line is I don't see that this scenario should result in unhandledrejection.

Finally ```module.export = SyncServer;``` in ```WebSocketSyncServer..js``` should be ```module.exports = SyncServer;``` at least to work in Node.js

**Comment by nponiros** [comment]
There is/was an issue where the server had to succeed during the first call in order for Syncable to work correctly. If the first call failed then there were many unhandled exceptions from within Dexie.Observable. Which versions of Dexie/Dexie.Observable/Dexie.Syncable are you using? Could you maybe clear the local database and try to do the first sync only after the server connected and then disconnect it and see if you still have issues? I am using the poll-pattern but it now works even when the server is disconnected (during first sync the server was connected).

On an other note, I don't know if you noticed in https://github.com/dfahlander/Dexie.js/issues/397 but I wrote a server to be used with the poll pattern and a database called NeDB. I was planing to add support for more databases and maybe for sockets. Perhaps we can work together on that.

**Comment by nevf** [comment]
@nponiros Thanks for the reply. Yes I'd already seen the ```Unhandled promise rejection: OpenFailedError: OpenFailedError``` errors on first sync of a new Dexie DB if no server is available and after reading the code comments understand the rationale for this. That said I'm not sure I agree with it.  More on that another time.

I deleted the db, started my server and did the first sync. If I kill the server I get ```Unhandled promise rejection: undefined``` in my ```window.onunhandledrejection``` handler. preceded by a Websocket ```ERR_CONNECTION_REFUSED``` error, which is to be expected. However again I don't want this unhandled rejection, which is caused by the same issue in my initial post. 

If I then refresh the Browser with the server not running I get the same unhandled rejection as above, which is the same as in my first post. 

In addition there is:
````
Unhandled promise rejection: TypeError: Cannot read property 'toString' of undefined
    at msg.reject (http://localhost:2015/jspm_packages/npm/dexie-observable@0.9.2/dist/dexie-observable.js:780:58)
````
which happens because ```error``` is undefined in:
````
db.sendMessage('response', { error: error.toString() }, msg.sender, { isFailure: true, requestId: msg.id });
````

Re. #397 yes I'd already seen your repo. It is great to see Dexie.Syncable getting some much needed attention. My interest is specifically using Websockets with Dexie.Syncable and MongoDB on the backend. If there is any overlap in what we are doing I'd certainly be interested in joining forces.

FYI I want to enable full offline use of my [Web App](http://myclibu.com), Clibu [web site](http://www.clibu.com)  which is why I'm looking closely at Dexie and Dexie.Syncable. This already uses Web sockets and RPC to work with Node.js and MongoDB on the server. I'd like to slot Dexie in between, eventually using Service Workers to turn Clibu into a Progressive Web App (PWA). 

I'm in Australia, hence the delay replying.

**Comment by nponiros** [comment]
Okay, now I understand your issue a bit better. The react pattern only has on onError, the poll pattern has multiple onError callbacks. So basically in the react pattern case, a promise is rejected multiple times which is not allowed (at least not by es6) and I'm guessing this causes the unhandledrejection. This is a bug I would say, at the very least we should not reject more than once. I'm not 100% sure as I can't test right now but that would be my guess.

Regarding the Observable exception: no clue why the function is even called. At least I didn't find anything in Syncable that would suggest that the function is called. Well a table is not the most suitable tool to browse the code so I might have missed something.

I will probably not find much time to look into the issues until the weekend but I will see what I can do.

Regarding a possible collaboration: okay so you basically already have a server and need the sync logic. You can have a look at the sync folder in my repo which basically is the code from the socket sample but made to work with the poll pattern and an async database. You can also look at the code in the db folder to see what tables I used and my db queries (NeDB has an API which is very similar to mongo db).

Well I'm in Germany so there is a bit of a time difference :)

**Comment by nevf** [comment]
Ok I've got to the bottom of the ```unhandledrejection``` issue. In ```WebSocketSyncProtocol.js``` the ```ws.onerror()``` function should _not_ call ```onError()``` 

From MDN [Writing WebSocket client applications](https://developer.mozilla.org/en-US/docs/Web/API/WebSockets_API/Writing_WebSocket_client_applications)
> If an error occurs while attempting to connect, first a simple event with the name "error" is sent to the WebSocket object (thereby invoking its onerror handler), and then the CloseEvent is sent to the WebSocket object (thereby invoking its onclose handler) to indicate the reason for the connection's closing.

The ```ws.onerror()``` call to ```onError()```  rejected the promise and then the subsequent ```ws.onclose()``` call to ```onError()``` did another reject, which caused the ```unhandledrejection```.

Further ```ws.onerror()``` should not call ```ws.close()``` as it is called by the Websocket code as per the quote above.

I've also updated the ```ws.onclose()``` ```ws.onerror()``` call to pass a customer Error object with the ```event.code``` which describes why the socket was closed. ```event.reason``` is undefined in Chrome in this specific issue.  See http://tools.ietf.org/html/rfc6455#section-7.4.1 for code info.

**Comment by nevf** [comment]
@nponiros Thanks, I'll definitely look at your NeDB sync code.

**Comment by nponiros** [comment]
@nevf Great that you solved the issue. If you have time maybe you could open a PR with your fixes to the sample code.

**Comment by nevf** [comment]
@nponiros Shall do. I haven't created a fork etc. yet. I'll probably wait until I get everything working to my satisfaction. I'm using your sync-server code and will post over there.
